### PR TITLE
Dev docker

### PR DESCRIPTION
**Pull request by pyoner**

### CODE REVIEW

**Comment by blokhin** [comment]
@knopki please, see our chat 🙌

File: package.json [hunk 1]
```diff
@@ -121,10 +124,13 @@
...
 		"debounce-promise": "^3.1.2",
+		"dotenv": "^16.0.3",
 		"optimade": "^2.1.0",
 		"optimade-mpds-nlp": "^0.1.3",
 		"sirv-cli": "^1.0.12",
 		"svelte-file-dropzone": "^0.0.15",
-		"tinycolor2": "^1.4.2"
+		"tinycolor2": "^1.4.2",
+		"znv": "^0.3.2",
+		"zod": "^3.20.2"
```
Comment (knopki) [inline package.json:134]: `dotenv`, `znv`, `zod` - they should be dev dependencies. They are required only in build time (or dev server?). Browser app (runtime) can't read environment variables or `.env` file.

Comment (pyoner) [reply package.json:134]: Yes, they should be dev deps.
Environment variables are injected in build time via https://github.com/pyoner/metis-gui/blob/dev-docker/package.json#L95
Ok, I'll fix it.

**Comment by knopki** [comment]
Removing the `METIS_RUNTIME_CONFIG` part looks like regression for me. With this PR it's useless for dev server - use can set configuration options from environment variables or `.env` file. But this settings override is build-only.

`METIS_RUNTIME_CONFIG` allows you to do a runtime configuration of a production build. Thus, if you have a docker image or just `index.html` hosted on any server, you can change e.g. the api url without rebuilding.

**Comment by knopki** [comment]
Also, `README.md` should be updated.

**Comment by pyoner** [comment]
> `METIS_RUNTIME_CONFIG` allows you to do a runtime configuration of a production build. Thus, if you have a docker image or just `index.html` hosted on any server, you can change e.g. the api url without rebuilding.

@knopki 
I understand you concern, but how often do you change environment variables on the production server?
If we inject env vars via `METIS_RUNTIME_CONFIG`, we are lose a validation process for env vars.

**Comment by knopki** [comment]
@pyoner 

> @knopki I understand you concern, but how often do you change environment variables on the production server?

I can't say for this project - I'm not owner of this project nor member of ops team. But in my experience in other projects - it is required very often and always pain-in-the-ass for (dev)ops. Cases:

- A/B testing (ok it's rarely API/auth only)
- same build for integration testing -> staging -> production environments (different urls)
- classic "debug in production"
- temporary deploys for every PR on services like Cloud Run where you don't know urls before deploy

However, none of this is relevant to this project 🤣 

In any case, this is just my opinion. I wanted to explain the motivation for adding this feature. That's why it's a comment, not a "request for changes".

> If we inject env vars via `METIS_RUNTIME_CONFIG`, we are lose a validation process for env vars.

Yep, that's bad.

File: package.json [hunk 2]
```diff
@@ -3,11 +3,11 @@
...
-		"build": "npm run clean && env BROWSERSLIST_CONFIG=./package.json rollup -c",
-		"build:light": "npm run clean && env BROWSERSLIST_CONFIG=./package.json rollup -c --light_mode",
-		"dev": "npm run clean && env BROWSERSLIST_CONFIG=./package.json rollup -c -w",
-		"dev:light": "npm run clean && env BROWSERSLIST_CONFIG=./package.json rollup -c -w --light_mode",
-		"start": "sirv dist --single",
+		"build": "npm run clean && env NODE_ENV=production BROWSERSLIST_CONFIG=./package.json rollup -c rollup.config.ts --configPlugin @rollup/plugin-typescript",
+		"build:light": "npm run clean && env NODE_ENV=production BROWSERSLIST_CONFIG=./package.json LIGHT_MODE=1 rollup -c rollup.config.ts --configPlugin @rollup/plugin-typescript",
+		"dev": "npm run clean && env NODE_ENV=development BROWSERSLIST_CONFIG=./package.json rollup -w -c rollup.config.ts --configPlugin @rollup/plugin-typescript",
+		"dev:light": "npm run clean && env NODE_ENV=development BROWSERSLIST_CONFIG=./package.json LIGHT_MODE=1 rollup -w -c rollup.config.ts --configPlugin @rollup/plugin-typescript",
+		"start": "sirv dist --single --host",
```
Comment (knopki) [inline package.json:10]: Maybe `sirv dist --single --host 0.0.0.0`?

Comment (pyoner) [reply package.json:10]: By default the host is 0.0.0.0 [see](https://github.com/lukeed/sirv/tree/master/packages/sirv-cli#network-access:~:text=Using%20%2D%2Dhost%20without%20a%20value%20is%20equivalent%20to%20%2D%2Dhost%200.0.0.0)

File: src/index.html [hunk 3]
```diff
@@ -18,7 +18,6 @@
 		<meta name="theme-color" content="#ffffff" />
 
 		<link rel="manifest" href="/build/public/manifest.json" />
-		<script>window._METIS_RUNTIME_CONFIG = {};</script>
```
Comment (blokhin) [inline src/index.html:21]: @pyoner is it okay, that we got rid of the runtime config feature? I think, `METIS_RUNTIME_CONFIG` was introduced by @knopki

Comment (pyoner) [reply src/index.html:21]: This commit **breaks** the compatibility with `METIS_RUNTIME_CONFIG`!
If you need this feature, I must revert the commit and write code to include the runtime config in the app.
But I don't see any case where this feature would be useful.

Comment (blokhin) [reply src/index.html:21]: @pyoner the point of `METIS_RUNTIME_CONFIG` is that we can configure the production build without the re-compilation. As for now, we can live without it, but it seems to be a comfortable feature :thinking:

Comment (pyoner) [reply src/index.html:21]: > @pyoner the point of `METIS_RUNTIME_CONFIG` is that we can configure the production build without the re-compilation. As for now, we can live without it, but it seems to be a comfortable feature thinking

ok, I'll fix it

Comment (blokhin) [reply src/index.html:21]: Thank you in advance! Much appreciated

**Comment by pyoner** [comment]
@blokhin @knopki 
this PR is not complete

I don't like injecting of the config file to the `index.html` file

I have an idea about `env` on runtime.
What if pass a **.env file** to the docker container, in the browser fetch this file, parse, validate and merge with compile-time **env file**
Or pass through the `docker-compose.yaml` file:
```yaml
environment:
 - METIS_<ENV_NAME>
```
And the docker endpoint script extract all envs that start with prefix `METIS_` and puts them in the `.env` file

**Comment by knopki** [comment]
@pyoner Second variant is nice. But maybe just dump to a json file, so you can skip parsing?

**Comment by knopki** [comment]
@pyoner But why? Library is not isomorphic (but may work, not sure), but [parsing is trivial](https://github.com/motdotla/dotenv/blob/master/lib/main.js#L11-L47). The method simply builds a javascript object. If config is already in JavaScriptObjectNotation, then ...

**Comment by pyoner** [comment]
> @pyoner But why? Library is not isomorphic (but may work, not sure), but [parsing is trivial](https://github.com/motdotla/dotenv/blob/master/lib/main.js#L11-L47). The method simply builds a javascript object. If config is already in JavaScriptObjectNotation, then ...

@knopki I think the library will work in the browser
Because `process.env` is a `Record<string, string | undefined>` and znv [parseEnv](https://github.com/lostfictions/znv/blob/master/src/parse-env.ts#L107-L108) function accepts envs in that format

**Comment by pyoner** [comment]
> @pyoner let's revive this PR. Do you use this new functionality? Does it work flawlessly for you?

@blokhin it works for me, but it is not completed, see https://github.com/basf/metis-gui/pull/120#issuecomment-1413592881

**Comment by blokhin** [comment]
Closing, but not deleting. Hope to return to this at some point later.

**Pull request closed without merging**